@vates/nbd-client/index.mjs

```diff
@@ -0,0 +1,224 @@
+    const answer = await this.#read(134)
+    this.#exportSize = answer.readBigUInt64BE(0)
+    const transmissionFlags = answer.readInt16BE(8)
+    assert.strictEqual(transmissionFlags & NBD_FLAG_HAS_FLAGS, NBD_FLAG_HAS_FLAGS, 'NBD_FLAG_HAS_FLAGS') // must always be 1 by the norm
+
+    // note : xapi server always send NBD_FLAG_READ_ONLY (3) as a flag
+  }
+
+  #read(length) {
+    return readChunkStrict(this.#serverSocket, length)
+  }
+
+  #write(buffer) {
+    return fromCallback.call(this.#serverSocket, 'write', buffer)
+  }
+
+  async #readInt32() {
+    const buffer = await this.#read(4)
+    return buffer.readInt32BE(0)
+  }
+
+  async #readInt64() {
+    const buffer = await this.#read(8)
+    return buffer.readBigUInt64BE(0)
+  }
+
+  #writeInt32(int) {
+    const buffer = Buffer.alloc(4)
+    buffer.writeInt32BE(int)
+    return this.#write(buffer)
+  }
+
+  async #readBlockResponse() {
+    // ensure at most one read occur in parallel
+    this.#nbCommands++
+    if (this.#waitingForResponse) {
+      return
+    }
+    this.#waitingForResponse = true
+    this.#nbCommands--
+    const magic = await this.#readInt32()
+
+    if (magic !== NBD_REPLY_MAGIC) {
+      throw new Error(`magic number for block answer is wrong : ${magic} ${NBD_REPLY_MAGIC}`)
+    }
+
+    const error = await this.#readInt32()
+    if (error !== 0) {
+      // @todo use error code from constants.mjs
+      throw new Error(`GOT ERROR CODE  : ${error}`)
+    }
+
+    const blockQueryId = await this.#readInt64()
+    const query = this.#commandQueryBacklog.get(blockQueryId)
+    if (!query) {
+      throw new Error(` no query associated with id ${blockQueryId}`)
+    }
+    this.#commandQueryBacklog.delete(blockQueryId)
+    const data = await this.#read(query.size)
+    query.resolve(data)
+
+    this.#waitingForResponse = false
+    if (this.#nbCommands > 0) {
+      await this.#readBlockResponse()
+    }
+  }
+
+  async readBlock(index, size = NBD_DEFAULT_BLOCK_SIZE) {
+    const queryId = this.#nextCommandQueryId
+    this.#nextCommandQueryId++
+
+    // create and send  command at once to ensure there is no concurrency issue
+    const buffer = Buffer.alloc(28)
+    buffer.writeInt32BE(NBD_REQUEST_MAGIC, 0) // it is a nbd request
+    buffer.writeInt16BE(0, 4) // no command flags for a simple block read
+    buffer.writeInt16BE(NBD_CMD_READ, 6) // we want to read a data block
+    buffer.writeBigUInt64BE(queryId, 8)
+    // byte offset in the raw disk
+    buffer.writeBigUInt64BE(BigInt(index) * BigInt(size), 16)
+    buffer.writeInt32BE(size, 24)
+
+    return new Promise((resolve, reject) => {
+      // this will handle one block response, but it can be another block
+      // since server does not guaranty to handle query in order
+      this.#commandQueryBacklog.set(queryId, {
+        size,
+        resolve,
+      })
+      // really send the command to the server
+      this.#write(buffer)
+        .then(() => this.#readBlockResponse())
```

Comment:
- don't wait for this method
- ensure this method never throw/reject
- in case of failure, reject all commands